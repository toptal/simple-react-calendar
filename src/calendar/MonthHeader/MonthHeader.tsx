import React, { FC, SyntheticEvent } from 'react'

import { IMonthHeaderButtonRenderProps, IMonthHeaderDayRenderProps, IMonthHeaderRenderProps } from '../../@types'
import { IGetDaysOfWeek } from '../helper'
import MonthHeaderButton from '../MonthHeaderButton'
import MonthHeaderDay from '../MonthHeaderDay'

export type Props = {
  activeMonth: Date
  blockClassName: string
  customRender?: IMonthHeaderRenderProps
  daysOfWeek: string[]
  getDaysOfWeek: (data: IGetDaysOfWeek) => string[]
  getMonthFormattedName: (day: Date) => string
  handleOnSwitchMonth: (event: SyntheticEvent<HTMLElement>) => void
  isMonthNextEnabled: boolean
  isMonthPrevEnabled: boolean
  renderMonthHeaderButton?: IMonthHeaderButtonRenderProps
  renderMonthHeaderDay?: IMonthHeaderDayRenderProps
  showDaysOfWeek: boolean
  showMonthName: boolean
  showMonthNext: boolean
  showMonthPrev: boolean
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6
}

const MonthHeader: FC<Props> = (props) => {
  const {
    activeMonth,
    blockClassName,
    customRender,
    daysOfWeek,
    getDaysOfWeek,
    getMonthFormattedName,
    handleOnSwitchMonth,
    isMonthNextEnabled,
    isMonthPrevEnabled,
    renderMonthHeaderButton,
    renderMonthHeaderDay,
    showDaysOfWeek,
    showMonthName,
    showMonthNext,
    showMonthPrev,
    weekStartsOn,
  } = props

  if (customRender) {
    return customRender(props)
  }

  if (!showMonthPrev && !showMonthName && !showMonthNext && !showDaysOfWeek) return null

  return (
    <header className={`${blockClassName}-month_header`}>
      <div className={`${blockClassName}-month_header_wrapper`}>
        {showMonthPrev && (
          <MonthHeaderButton
            blockClassName={blockClassName}
            customRender={renderMonthHeaderButton}
            enabled={isMonthPrevEnabled}
            onClick={handleOnSwitchMonth}
            type="prev"
          />
        )}
        {showMonthName && (
          <h3 className={`${blockClassName}-month_header_title`}>{getMonthFormattedName(activeMonth)}</h3>
        )}
        {showMonthNext && (
          <MonthHeaderButton
            blockClassName={blockClassName}
            customRender={renderMonthHeaderButton}
            enabled={isMonthNextEnabled}
            onClick={handleOnSwitchMonth}
            type="next"
          />
        )}
      </div>
      <div className={`${blockClassName}-month_header_days`}>
        {showDaysOfWeek &&
          getDaysOfWeek({dayIndex: weekStartsOn, daysOfWeek}).map((day, index) => (
            <MonthHeaderDay
              blockClassName={blockClassName}
              customRender={renderMonthHeaderDay}
              day={day}
              isWeekend={index >= weekStartsOn + 5}
              key={day}
            />
          ))}
      </div>
    </header>
  )
}

MonthHeader.displayName = 'MonthHeader'

export default MonthHeader
