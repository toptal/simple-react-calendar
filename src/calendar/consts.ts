export const BLOCK_CLASS_NAME = 'calendar'
export const NEXT_MONTH_TITLE = 'Next month'
export const PREV_MONTH_TITLE = 'Previous month'
export const DAYS_IN_WEEK = [0, 1, 2, 3, 4, 5, 6]
export const DAYS_OF_WEEK = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']
