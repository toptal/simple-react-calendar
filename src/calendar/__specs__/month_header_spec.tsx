import { shallow } from 'enzyme'
import React from 'react'

import { NEXT_MONTH_TITLE, PREV_MONTH_TITLE } from '../consts'
import MonthHeader from '../month_header'

describe('MonthHeader', () => {
  let instance, props, wrapper

  beforeEach(() => {
    props = getProps()
    wrapper = shallow(<MonthHeader {...props} />)
    instance = wrapper.instance()
  })

  describe('#_switchMonth', () => {
    it('calls #onMonthChange', () => {
      wrapper.setProps({ maxDate: new Date(2015, 7, 1) })
      instance._switchMonth(1)

      expect(props.onMonthChange).toHaveBeenCalledTimes(1)

      expect(props.onMonthChange).toBeCalledWith(new Date(2015, 8, 17))
    })
  })

  describe('#render', () => {
    it('renders <MonthHeader />', () => {
      expect(wrapper).toMatchSnapshot()
    })

    describe('when prop `minDate` is after prop `activeMonth`', () => {
      it('renders <MonthHeader />', () => {
        wrapper.setProps({ minDate: new Date(2015, 8, 17) })

        expect(wrapper).toMatchSnapshot()
      })
    })

    describe('when prop `maxDate` is before prop `activeMonth`', () => {
      it('renders <MonthHeader />', () => {
        wrapper.setProps({ maxDate: new Date(2015, 6, 17) })

        expect(wrapper).toMatchSnapshot()
      })
    })
  })
})
const getProps = (overrides = {}) => ({
  activeMonth: new Date(2015, 7, 17),
  blockClassName: 'example-class',
  headerNextTitle: NEXT_MONTH_TITLE,
  headerPrevTitle: PREV_MONTH_TITLE,
  onMonthChange: jest.fn(),
  ...overrides
})
