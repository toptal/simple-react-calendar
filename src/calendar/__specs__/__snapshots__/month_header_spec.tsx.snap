// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthHeader #render renders <MonthHeader /> 1`] = `
<div
  className="example-class-month_header"
>
  <HeaderButton
    blockClassName="example-class"
    enabled={true}
    onClick={[Function]}
    title="Previous month"
    type="prev"
  />
  <div
    className="example-class-month_header_title"
  >
    August 2015
  </div>
  <HeaderButton
    blockClassName="example-class"
    enabled={true}
    onClick={[Function]}
    title="Next month"
    type="next"
  />
</div>
`;

exports[`MonthHeader #render when prop \`maxDate\` is before prop \`activeMonth\` renders <MonthHeader /> 1`] = `
<div
  className="example-class-month_header"
>
  <HeaderButton
    blockClassName="example-class"
    enabled={true}
    onClick={[Function]}
    title="Previous month"
    type="prev"
  />
  <div
    className="example-class-month_header_title"
  >
    August 2015
  </div>
  <HeaderButton
    blockClassName="example-class"
    enabled={false}
    onClick={[Function]}
    title="Next month"
    type="next"
  />
</div>
`;

exports[`MonthHeader #render when prop \`minDate\` is after prop \`activeMonth\` renders <MonthHeader /> 1`] = `
<div
  className="example-class-month_header"
>
  <HeaderButton
    blockClassName="example-class"
    enabled={false}
    onClick={[Function]}
    title="Previous month"
    type="prev"
  />
  <div
    className="example-class-month_header_title"
  >
    August 2015
  </div>
  <HeaderButton
    blockClassName="example-class"
    enabled={true}
    onClick={[Function]}
    title="Next month"
    type="next"
  />
</div>
`;
